endless-sky/endless-sky
# Make the mission dialog "Decline" button optional and allow Ok/Cancel instead of Accept/Decline

**Feature:** This PR implements the feature request detailed and discussed in issue #6632

## Feature Details
The "Decline" button in mission dialogs only appears if a `to decline` condition is met. If it is *not* met, you only get an "Ok" button, like the built-in dialogs.

```ruby
        dialog
            "Here's some nice text"
            # Only show the "Decline" button if this condition is met:
            to decline
                "some condition" == 123
            "Some more text"

        dialog "Here's some nice text"
            # Only show the "Decline" button if this condition is met:
            to decline
                "some condition" == 123

        dialog phrase "some phrase name"
            # Only show the "Decline" button if this condition is met:
            to decline
                "some condition" == 123

        dialog
            phrase
                 phrase
                       "some phrase name"
            # Only show the "Decline" button if this condition is met:
            to decline
                "some condition" == 123

        dialog
            # Only show the "Decline" button if this condition is met:
            to decline
                "some condition" == 123
            phrase
                 phrase
                       "some phrase name"
```

## UI Screenshots
![destroy-earth](https://github.com/endless-sky/endless-sky/assets/116329264/0dfa3066-13e5-4f5f-949a-6dedb2af8394)

## Usage Examples

For missions with a short dialog that must not be able to decline the mission, you don't have to use a conversation. One example is `Remnant: Cognizance 10`

## Testing Done
1. Try lots of dialogs in the game and make sure they work as intended.
2. Put this in data/[todeclinetest.txt](https://github.com/endless-sky/endless-sky/files/12448867/todeclinetest.txt)
3. Visit Earth to choose whether dialogs on Mars should have "Decline" buttons or not.
4. Fly to Mars to see the dialogs. There should be four of them: two with bands (from phrases) and two telling you to fly back to Earth. Instead of Accept/Decline, you'll see Ok/Cancel, if everything went according to plan.
5. Repeat steps 3-5 ad nauseam.

### Automated Tests Added
None. I can't think of any way to test this automatically.

## Performance Impact
N/A
